File: nqueen.py
def solveNQueens(N):
    def backtrack(row, columns, diagonals1, diagonals2, board):
        if row == N:
            result.append(board)
            return
        
        for col in range(N):
            if col in columns or (row - col) in diagonals1 or (row + col) in diagonals2:
                continue
            
            # Place the queen
            columns.add(col)
            diagonals1.add(row - col)
            diagonals2.add(row + col)
            new_board = board + ["." * col + "Q" + "." * (N - col - 1)]
            
            backtrack(row + 1, columns, diagonals1, diagonals2, new_board)
            
            # Remove the queen (backtrack)
            columns.remove(col)
            diagonals1.remove(row - col)
            diagonals2.remove(row + col)

    result = []
    backtrack(0, set(), set(), set(), [])
    return result

# Example Usage
N = 4
solutions = solveNQueens(N)
for solution in solutions:
    for row in solution:
        print(row)
    print()  # Print a blank line between solutions

File: N_Queen.cpp
#include <iostream>
#include <vector>
using namespace std;

bool isSafe(const vector<vector<int>>& board, int row, int col, int N) {
    for (int i = 0; i < col; i++)
        if (board[row][i]) return false;
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j]) return false;
    for (int i = row, j = col; i < N && j >= 0; i++, j--)
        if (board[i][j]) return false;
    return true;
}

bool solveNQueensUtil(vector<vector<int>>& board, int col, int N) {
    if (col >= N) return true;
    for (int i = 0; i < N; i++) {
        if (isSafe(board, i, col, N)) {
            board[i][col] = 1;
            if (solveNQueensUtil(board, col + 1, N)) return true;
            board[i][col] = 0;
        }
    }
    return false;
}

void solveNQueens(int N) {
    vector<vector<int>> board(N, vector<int>(N, 0));
    if (solveNQueensUtil(board, 0, N)) {
        for (const auto& row : board) {
            for (int cell : row) cout << (cell ? "Q " : ". ");
            cout << endl;
        }
    } else {
        cout << "Solution does not exist" << endl;
    }
}

int main() {
    int N = 4; // Change N to solve for different sizes
    solveNQueens(N);
    return 0;
}
